python/lsst/cp/pipe/defects.py

```diff
@@ -0,0 +1,645 @@
+        detNum = dataRef.dataId[self.config.ccdKey]
+
+        if self.config.mode == "MASTER":
+            filename = f"defectPlot_det{detNum}_master-{imageType}_for-exp{visit}.pdf"
+        elif self.config.mode == "VISITS":
+            filename = f"defectPlot_det{detNum}_{imageType}_exp{visit}.pdf"
+
+        filenameFull = os.path.join(dirname, filename)
+        self.log.info(f'Wrote amp histogram to {filenameFull}')
+
+        with warnings.catch_warnings():
+            msg = "Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure."
+            warnings.filterwarnings("ignore", message=msg)
+            with PdfPages(filenameFull) as pdfPages:
+                self._plotAmpHistogram(dataRef, expCopy, visit, nSig)  # NB: modifies exp to scale by expTime
+                pdfPages.savefig()
+
+                self._plotDefects(exp, visit, defects, imageType)
+                pdfPages.savefig()
+
+    def _plotDefects(self, exp, visit, defects, imageType):  # pragma: no cover
+        """TODO: Write docstring
+        """
+        expCopy = exp.clone()
+        del exp  # for safety - no longer needed as we have a copy
+        maskedIm = expCopy.maskedImage
+
+        defects.maskPixels(expCopy.maskedImage, "BAD")
+        detector = expCopy.getDetector()
+
+        disp = afwDisplay.Display(0, reopenPlot=True, dpi=200)  # , dpi=200)
+
+        if imageType == "flat":  # set each amp image to have a mean of 1.00
+            for amp in detector:
+                ampIm = maskedIm.image[amp.getBBox()]
+                ampIm -= afwMath.makeStatistics(ampIm, afwMath.MEANCLIP).getValue() + 1
+
+        mpDict = maskedIm.mask.getMaskPlaneDict()
+        for plane in mpDict.keys():
+            if plane in ['BAD']:
+                continue
+            disp.setMaskPlaneColor(plane, afwDisplay.IGNORE)
+
+        # if self._getNumGoodPixels(maskedIm) >= 1:
+        disp.scale('asinh', 'zscale')
+        disp.setMaskTransparency(80)
+        disp.setMaskPlaneColor("BAD", afwDisplay.RED)
+
+        disp.setImageColormap('gray')
+        title = f"Detector: {detector.getName()[-3:]} {detector.getSerial()}"
+        title += f", Type: {imageType}, visit: {visit}"
+        disp.mtv(maskedIm, title=title)
+
+        cgUtils.overlayCcdBoxes(detector, isTrimmed=True, display=disp)
+
+    def _plotAmpHistogram(self, dataRef, exp, visit, nSigmaUsed):  # pragma: no cover
+        """
+        Make a histogram of the distribution of pixel values for each amp.
+
+        Main image data histogram plotted in blue. Edge pixels, if masked, are
+        in red. Note that masked edge pixels do not contribute to the
+        underflow and overflow numbers.
+
+        Parameters
+        ----------
+        pdfPages: `matplotlib.backends.backend_pdf.PdfPages`
+            If not None, append per-amp histograms to this object
+        """
+        fig, ax = plt.subplots(nrows=4, ncols=4, sharex='col', sharey='row', figsize=(13, 10))
```

Comment:
4 rows and 4 columns implies that this only works on LSST data, no (2x8 amps)? Would it be better to make it scale to the number of amps?

Reply:
It would, but it's not immediately obvious to me how to do that in a robust but still pretty way. For now I'll just make it so that it explicitly only plots LSST stuff I think, and if people from other cameras end up caring we can worry about that when we get to that bridge.